src/internal/pjsdb/sanitize.go

```diff
@@ -0,0 +1,64 @@
+package pjsdb
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+)
+
+// sanitize.go contains a sanitization pattern to decouple the CRUD API from *fileset.Storage and validate
+// CreateJob() parameters.
+
+// CreateJobRequest is a bundle of related fields required for a CreateJob() invocation.
+// In pfsdb, we used a pattern of forcing the caller to convert their resources to a _Info struct, but its problematic since
+// each database function only really needs a subset of the fields and it is not clear to the caller which fields are required.
+// CreateJobRequest cannot be used directly, it has to be first sanitized via Sanitize()
+type CreateJobRequest struct {
+	Parent      JobID
+	Input, Spec string
+}
+
+type createJobRequest struct {
+	parent      sql.NullInt64
+	input, spec string
+	specHash    []byte
+}
+
+// Sanitize should be invoked prior to calling CreateJob() to ensure that a job can be properly created.
+// this functionally is separate from the CreateJob() so that CreateJob() doesn't require access to a *fileset.Storage
+func (req CreateJobRequest) Sanitize(ctx context.Context, tx *pachsql.Tx, s *fileset.Storage) (emptyReq createJobRequest, err error) {
```

Comment:
It’s not great for a public method to return a private type.

Reply:
I agree.  I'm open to other alternatives if they can achieve the same kind of restrictions. This is my workaround for preventing a caller from calling CreateJob() without first calling Sanitize(). You could just call Sanitize() within CreateJob(), but then CreateJob() would depend on *fileset.Storage().